src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -3359,53 +3363,85 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
-                  // infer.checkNames must not use UnitType: it may not be a valid assignment, or the setter may return another type from Unit
-                  // TODO: just make it an error to refer to a non-existent named arg, as it's far more likely to be
-                  //       a typo than an assignment passed as an argument
-                  case AssignOrNamedArg(lhs@Ident(name), rhs) =>
-                    // named args: only type the righthand sides ("unknown identifier" errors otherwise)
-                    // the assign is untyped; that's ok because we call doTypedApply
-                    typedArg0(rhs) match {
-                      case (rhsTyped, tp) => (treeCopy.AssignOrNamedArg(arg, lhs, rhsTyped), NamedType(name, tp))
-                    }
-                  case treeInfo.WildcardStarArg(_) =>
-                    typedArg0(arg) match {
-                      case (argTyped, tp) => (argTyped, RepeatedType(tp))
-                    }
-                  case _ =>
-                    typedArg0(arg)
-                }
-              }.unzip
+            val (args1, argTpes, maybeRetypeArgs) = context.savingUndeterminedTypeParams() {
+              val amode = forArgMode(fun, mode)
+              map2(args, altArgPts) { (arg, argPtAlts) => typedSingleArg(arg, argPtAlts, amode, matchAsPartial = true) }.unzip3
             }
             if (context.reporter.hasErrors)
               setError(tree)
             else {
               inferMethodAlternative(fun, undetparams, argTpes, pt)
-              doTypedApply(tree, adaptAfterOverloadResolution(fun, mode.forFunMode, WildcardType), args1, mode, pt)
+              val args2 = if(maybeRetypeArgs contains true) {
+                // There were pattern-matching anonymous functions in the argument list which got typed as PartialFunction
+                // but if the inferred alternative does not actually need a PartialFunction we retype it as Function1.
+                def paramTypes(tpe: Type): List[Type] = tpe match {
+                  case ExistentialType(_, qtpe) => paramTypes(qtpe)
+                  case MethodType(syms, _) => syms.map(_.tpe)
+                  case TypeRef(_, _, tpes) => tpes.init
+                  case _ => Nil
+                }
+                val selectedParams = paramTypes(normalize(fun.tpe))
+                if(selectedParams.nonEmpty) {
+                  context.savingUndeterminedTypeParams() {
+                    val amode = forArgMode(fun, mode)
+                    val keepTypes = map3(selectedParams, maybeRetypeArgs, args1) { (param, maybeRetype, arg1) =>
+                      if (maybeRetype && !isPartialFunctionType(param)) None else Some(arg1)
+                    }
+                    map3(args, altArgPts, keepTypes) {
+                      case (_, _, Some(keep)) => keep
+                      case (arg, argPtAlts, None) => typedSingleArg(arg, argPtAlts, amode, matchAsPartial = false)._1
```

Comment:
@retronym wdyt about throwing away a type checked tree and typing it again? Are you aware of other places where we do this? Any potential problems that come to mind with side-effects of type checking, for examples related to definitions within that tree? I cannot think of any from the top of my head.

The alternative would be to generate a PartialFunction instance in case an overloaded method has alternatives with `Function1` and `PF` parameter types, and the `Function1` alternative is chosen.